packages/material-ui-lab/src/Autocomplete/Autocomplete.js

```diff
@@ -185,6 +185,11 @@ const Autocomplete = React.forwardRef(function Autocomplete(props, ref) {
     renderInput,
     renderOption: renderOptionProp,
     renderTags,
+    titles = {
```

Comment:
Is "titles" the right prop name to use for this? I appreciate that they're for tooltips, and the label prop-name in tooltips is "title" (for better or worse), but should that carry through to this prop?

Reply:
I followed the issue for this naming. Is it better to use `customTooltipTitles` or something like this?

Reply:
Something like `tooltipTitles` would be better, but I haven’t put any thought into this.

Reply:
IMHO, `tooltipTitles` is more intuitive. If others agree, I'll make the change.

Reply:
I wish we had components that already solve this problem, but we don't, as far as I know. `tooltipTitles` sounds already better. What about `i18n`?
@dmtrKovalenko Did you have to solve a similar problem with the picker?

Reply:
Yeah, actually I prefer a way to send several separate props, that allowes to pass components. Like `<Translate key="blablabla" />`.

Reply:
@dmtrKovalenko Thanks. In this case, I think that we should only support a string. I don't think that we should pull-in the Tooltip by default. Regarding the x props vs one prop. I share a possible advantage of the latter: https://github.com/mui-org/material-ui/issues/18202#issuecomment-549830562

Reply:
So, am I good to go with `tooltipTitles`?

Reply:
@soltanloo Hold on, let's benchmark how this problem was solved in the past by people:

- http://jquense.github.io/react-widgets/api/Combobox/#messages => `messages`
- https://reactdatepicker.com/#example-today-button => `todayButton`
- https://ant.design/components/config-provider/ => `locale`
- https://ej2.syncfusion.com/react/documentation/common/localization/ => `L10n`
- https://ag-grid.com/javascript-grid-internationalisation/ => `localeText `
- https://blueprintjs.com/docs/#datetime/dateinput.localization => `todayButtonText `
- https://www.telerik.com/kendo-react-ui/components/intl/l10n/loading-messages/ => `messages`
- https://devexpress.github.io/devextreme-reactive/react/grid/docs/guides/localization/ => `messages`

Oh, I was wrong, I had forgotten about our `labelRowsPerPage `, and `noOptionsText` and `loadingText` props. It seems that the most advanced and complex components we are going to build, the more visible this internationalization problem is going to be visible: https://github.com/telerik/kendo-react-messages/blob/master/messages/en-US/en-US.json. I think that it's time to think of a global solution to the problem.

Reply:
What about:

- Material-UI provides translations for the most common locales, outside of the box. We can focus on 5 locales (BCP 47 language tag) to start with, the five we server-side render the documentation with (en-US, zh-CN, ru-RU, pr-BR, es-ES). We let the community contribute more.
- We use the en-US translations by default.
- We provide a way to customize each component translation messages: a `messages` prop. If not all the keys are provided, we fallback to the default values.
- We provide a way to use the context to apply the translation messages:
`themes.props.MuiX.messages`

```jsx
import { createMuiTheme, enUS, frFR } from '@material-ui/core';

const theme = createMuiTheme({}, locale === 'en' ? enUS : frFR);
```

*packages/material-ui/src/messages/en-US.js*

```js

const messages = {
  props: {
    MuiTablePagination: {
      labelRowsPerPage: undefined, // undefined triggers the default value
    },
    MuiAutocomplete: {
      message: undefined,
    },
  },
};

export default messages;
```

*packages/material-ui/src/messages/fr-FR.js*

```js
import tablePaginationMessages from './TablePagination';
import autocompleteMessages from './Autocomplete';

const messages = {
  props: {
    MuiTablePagination: {
      labelRowsPerPage: 'Résultats par page :'
    },
    MuiAutocomplete: {
      message: {
        loading: 'Chargement…',
        noOptions: 'Pas de résultats',
        clear: 'Vider',
        closePopup: 'Ouvrir',
        openPopup: 'Fermer',
      },
    },
  },
};

export default messages;
```

Reply:
It is a nice idea, but (just a proposal) maybe it will be easier to write translations if object keys will be the actual message like

```
 MuiTablePagination: {
   'Results per page': 'Résultats par page :'
 },
```

Moreover then it will be possible to provide a codemod, or some cli generator that will generate such localization files for specific components with the default translation?

Reply:
@dmtrKovalenko Haha, the eternal [tradeoff](https://dev.to/omaiboroda/three-ways-to-name-i18n-translation-keys-2fed) on the natural vs code translation keys ✌️. What about the case where we accept a function?

Reply:
I've worked with simple workarounds for this localization issue. I'd be glad to contribute in this issue. 🙂